python_modules/libraries/dagster-openai/dagster_openai/resources.py

```diff
@@ -0,0 +1,382 @@
+        }
+        if hasattr(usage, "completion_tokens"):
+            usage_metadata["openai.completion_tokens"] = usage.completion_tokens
+        _add_to_asset_metadata(context, usage_metadata, output_name)
+
+        return response
+
+    return wrapper
+
+
+@experimental
+class OpenAIResource(ConfigurableResource):
+    """This resource is wrapper over the
+    `openai library <https://github.com/openai/openai-python>`_.
+
+    By configuring this OpenAI resource, you can interact with OpenAI API
+    and log its usage metadata in the asset metadata.
+
+    Examples:
+        .. code-block:: python
+
+            import os
+
+            from dagster import AssetExecutionContext, Definitions, EnvVar, asset, define_asset_job
+            from dagster_openai import OpenAIResource
+
+
+            @asset(compute_kind="OpenAI")
+            def openai_asset(context: AssetExecutionContext, openai: OpenAIResource):
+                with openai.get_client(context) as client:
+                    client.chat.completions.create(
+                        model="gpt-3.5-turbo",
+                        messages=[{"role": "user", "content": "Say this is a test"}]
+                    )
+
+            openai_asset_job = define_asset_job(name="openai_asset_job", selection="openai_asset")
+
+            defs = Definitions(
+                assets=[openai_asset],
+                jobs=[openai_asset_job],
+                resources={
+                    "openai": OpenAIResource(api_key=EnvVar("OPENAI_API_KEY")),
+                },
+            )
+    """
+
+    api_key: str = Field(description=("OpenAI API key. See https://platform.openai.com/api-keys"))
+
+    _client: Client = PrivateAttr()
+
+    @classmethod
+    def _is_dagster_maintained(cls) -> bool:
+        return True
+
+    def _wrap_with_usage_metadata(
+        self,
+        api_endpoint_class: ApiEndpointClassesEnum,
+        context: AssetExecutionContext,
+        output_name: Optional[str],
+    ):
+        for attribute_names in API_ENDPOINT_CLASSES_TO_ENDPOINT_METHODS_MAPPING[api_endpoint_class]:
+            curr = self._client.__getattribute__(api_endpoint_class.value)
+            # Get the second to last attribute from the attribute list to reach the method.
+            i = 0
+            while i < len(attribute_names) - 1:
+                curr = curr.__getattribute__(attribute_names[i])
+                i += 1
+            # Wrap the method.
+            curr.__setattr__(
+                attribute_names[i],
+                with_usage_metadata(
+                    context=context,
+                    output_name=output_name,
+                    func=curr.__getattribute__(attribute_names[i]),
+                ),
+            )
+
+    def setup_for_execution(self, context: InitResourceContext) -> None:
+        # Set up an OpenAI client based on the API key.
+        self._client = Client(api_key=self.api_key)
+
+    @contextmanager
```

Comment:
probably need a `@public` decorator here to make it show up in doc string. we can add and test that in the docs PR tho.